lib/web/websocket/connection.js

```diff
@@ -214,13 +216,80 @@ function establishWebSocketConnection (url, protocols, client, handler, options)
 }
 
 /**
- * @param {import('./websocket').Handler} handler
- * @param {number} code
- * @param {any} reason
- * @param {number} reasonByteLength
+ * @see https://whatpr.org/websockets/48.html#close-the-websocket
+ * @param {import('./websocket').Handler} object
+ * @param {number} [code]
+ * @param {string} [reason]
  */
-function closeWebSocketConnection (handler, code, reason, reasonByteLength) {
-  handler.onClose(code, reason, reasonByteLength)
+function closeWebSocketConnection (object, code, reason, validate = false) {
+  // 1. If code was not supplied, let code be null.
+  code ??= null
+
+  // 2. If reason was not supplied, let reason be the empty string.
+  reason ??= ''
+
+  // 3. Validate close code and reason with code and reason.
+  if (validate) validateCloseCodeAndReason(code, reason)
+
+  // 4. Run the first matching steps from the following list:
+  //     - If object’s ready state is CLOSING (2) or CLOSED (3)
+  //     - If the WebSocket connection is not yet established [WSP]
+  //     - If the WebSocket closing handshake has not yet been started [WSP]
+  //     - Otherwise
+  if (isClosed(object.readyState) || isClosing(object.readyState)) {
+    // Do nothing.
+  } else if (!isEstablished(object.readyState)) {
+    // Fail the WebSocket connection and set object’s ready state to CLOSING (2). [WSP]
+    failWebsocketConnection(object)
+    object.readyState = states.CLOSING
+  } else if (object.closeState === sentCloseFrameState.NOT_SENT) {
+    // Start the WebSocket closing handshake and set object’s ready state to CLOSING (2). [WSP]
+    object.closeState = sentCloseFrameState.PROCESSING
+
+    const frame = new WebsocketFrameSend()
+
+    // If neither code nor reason is present, the WebSocket Close
+    // message must not have a body.
+
+    // If code is present, then the status code to use in the
+    // WebSocket Close message must be the integer given by code.
+    // If code is null and reason is the empty string, the WebSocket Close frame must not have a body.
+    // If reason is non-empty but code is null, then set code to 1000 ("Normal Closure").
+    if (reason.length !== 0 && code === null) {
+      code = 1000
+    }
+
+    // If code is set, then the status code to use in the WebSocket Close frame must be the integer given by code.
+    assert(code === null || Number.isInteger(code))
+
```

Comment:
why assert if we had before set code to 1000?

```suggestion
    } else {
      // If code is set, then the status code to use in the WebSocket Close frame must be the integer given by code.
      assert(code === null || Number.isInteger(code))
    }
```

Reply:
we only set it to 1000 if code is null and reason is non-empty